tsl/src/nodes/columnar_scan/columnar_scan.c

```diff
@@ -83,12 +86,68 @@ typedef struct QualProcessState
 	List *vectorized_quals;
 	List *nonvectorized_quals;
 
+	/*
+	 * Further split nonvectorized quals into quals on segmentby columns and
+	 * non-segmentby columns.
+	 */
+	List *segmentby_quals;
+	List *nonsegmentby_quals;
+
 	/* Scankeys created from scankey_quals */
 	ScanKey scankeys;
 	unsigned int scankeys_capacity;
 	unsigned int nscankeys;
+
+	/* Scratch area for processing */
+	bool relvar_found;
 } QualProcessState;
 
+static bool
+segmentby_qual_walker(Node *qual, QualProcessState *qpc)
+{
+	if (qual == NULL)
+		return false;
```

Comment:
Is it possible to trigger this case in a test?

Reply:
I don't think so. We only reach here if a non-segmentby filter is found. But all filters on other columns are vectorized, so never parsed here. Currently, there is no way to turn off vectorized filters.

Reply:
Actually, we should reach here if segmentby filter is found, so something is off. Let me investigate.

Reply:
Ok, so I think I know more what is going on.

This is a base case in the pattern recommended for expression walkers in PG. It only reaches the base case if it descends down a child node which is NULL. For the expressions we evaluate, there are mostly no child nodes at the leaves, e.g., a Var or Const does not have children so the walker does not get to this point.